pkg/controller/instance/reconcile_operation.go

```diff
@@ -0,0 +1,476 @@
+package instance
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"log"
+	"reflect"
+
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/uuid"
+	"k8s.io/client-go/tools/record"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/kudobuilder/kudo/pkg/apis/kudo/v1beta1"
+	kudov1beta1 "github.com/kudobuilder/kudo/pkg/apis/kudo/v1beta1"
+	"github.com/kudobuilder/kudo/pkg/apitools/instance"
+	"github.com/kudobuilder/kudo/pkg/engine"
+	"github.com/kudobuilder/kudo/pkg/util/kudo"
+)
+
+const (
+	instanceCleanupFinalizerName = "kudo.dev.instance.cleanup"
+	snapshotAnnotation           = "kudo.dev/last-applied-instance-state"
+)
+
+type reconcileOperation struct {
```

Comment:
In theory, a struct with methods provides some meaningful encapsulation and hides implementation details. But in this case, it just obfuscates the controller code making it harder to grok. I don't think `reconcileOperation` is needed at all. I'd rather keep the abstraction flat and have a bunch of helper methods.